include/boost/hana/sort.hpp

```diff
@@ -68,82 +68,171 @@ BOOST_HANA_NAMESPACE_BEGIN
-        struct insertion_sort_impl<Pred, std::index_sequence<Result1, Result...>, T, Ts...> {
-            using type = typename insertion_sort_impl<
+        template <
+            typename Pred,
+            std::size_t l0,
+            std::size_t ...l,
+            std::size_t r0,
+            std::size_t ...r>
+        struct merge_helper<
+            Pred,
+            std::index_sequence<l0, l...>,
+            std::index_sequence<r0, r...>
+        > {
+            using type = typename merge<
                 Pred,
-                typename insert<
-                    Pred, T, (bool)Pred::template apply<T, Result1>::value,
-                    std::index_sequence<>,
-                    Result1, Result...
-                >::type,
-                Ts...
+                (bool)Pred::template apply<r0, l0>::value,
+                std::index_sequence<l0, l...>,
+                std::index_sequence<r0, r...>
             >::type;
         };
 
-        template <typename Pred, std::size_t T, std::size_t ...Ts>
-        struct insertion_sort_impl<Pred, std::index_sequence<>, T, Ts...> {
-            using type = typename insertion_sort_impl<
-                Pred, std::index_sequence<T>, Ts...
-            >::type;
+        template <std::size_t Nr, typename Left, typename Right, typename=void>
+        struct split;
+
+        template <
+            std::size_t Nr,
+            std::size_t ...l,
+            std::size_t ...r,
+            std::size_t r0>
+        struct split<
+            Nr,
+            std::index_sequence<l...>,
+            std::index_sequence<r0 ,r...>,
+            typename std::enable_if<Nr!=0>::type
+        > {
+            using sp = split<
+                Nr-1,
+                std::index_sequence<l..., r0>,
+                std::index_sequence<r...>
+            >;
+            using left = typename sp::left;
+            using right = typename sp::right;
         };
 
-        template <typename Pred, typename Result>
-        struct insertion_sort_impl<Pred, Result> {
-            using type = Result;
+        template <std::size_t ...l, std::size_t ...r>
+        struct split<0, std::index_sequence<l...>, std::index_sequence<r...>> {
+            using left = std::index_sequence<l...>;
+            using right = std::index_sequence<r...>;
         };
 
-        template <typename Pred, typename Indices>
-        struct sort_helper;
+        template <typename Pred, typename Sequence>
+        struct merge_sort_impl;
 
-        template <typename Pred, std::size_t ...i>
-        struct sort_helper<Pred, std::index_sequence<i...>> {
-            using type = typename insertion_sort_impl<
-                Pred, std::index_sequence<>, i...
+        template <typename Pred, std::size_t ...seq>
+        struct merge_sort_impl<Pred, std::index_sequence<seq...>> {
+            using sequence = std::index_sequence<seq...>;
+            using sp = split<
+                sequence::size() / 2,
```

Comment:
Why not use `sizeof...(seq)`?

Reply:
I can't remember a specific reason for this, but using `sequence::size()`, the `using sp =` statement only depends on `sequence`.

Do you see any reason I should use `sizeof` ?